cmd/crossplane/main.go

```diff
@@ -17,176 +17,121 @@ limitations under the License.
-		mgr := setupManager(log, syncPeriod)
+		log := logging.NewLogrLogger(zl.WithName("crossplane"))
+		log.Debug("Starting", "sync-period", syncPeriod.String())
+
+		cfg, err := ctrl.GetConfig()
+		kingpin.FatalIfError(err, "Cannot get config")
+
+		mgr, err := ctrl.NewManager(cfg, ctrl.Options{SyncPeriod: syncPeriod})
+		kingpin.FatalIfError(err, "Cannot create manager")
 
-		log.Info("Adding controllers")
-		kingpin.FatalIfError(controllerSetupWithManager(mgr), "Cannot add controllers to manager")
+		kingpin.FatalIfError(apis.AddToScheme(mgr.GetScheme()), "Cannot add core Crossplane APIs to scheme")
+		kingpin.FatalIfError(workload.Setup(mgr, log), "Cannot setup workload controllers")
+		kingpin.FatalIfError(mgr.Start(ctrl.SetupSignalHandler()), "Cannot start controller manager")
 
-		// Start the Cmd
-		log.Info("Starting the manager")
-		kingpin.FatalIfError(mgr.Start(signals.SetupSignalHandler()), "Cannot start controller")
 	case extManageCmd.FullCommand():
-		// the "stacks manage" command is being run, the only controllers we should add to the
-		// manager are the stacks controllers
+		log := logging.NewLogrLogger(zl.WithName("stack-manager"))
+		log.Debug("Starting", "sync-period", syncPeriod.String())
 
-		mgr := setupManager(log, syncPeriod)
+		cfg, err := ctrl.GetConfig()
+		kingpin.FatalIfError(err, "Cannot get config")
 
-		log.Info("Adding controllers")
-		kingpin.FatalIfError(stacksControllerSetupWithManager(mgr), "Cannot add controllers to manager")
+		mgr, err := ctrl.NewManager(cfg, ctrl.Options{SyncPeriod: syncPeriod})
+		kingpin.FatalIfError(err, "Cannot create manager")
+
+		kingpin.FatalIfError(apis.AddToScheme(mgr.GetScheme()), "Cannot add core Crossplane APIs to scheme")
+		kingpin.FatalIfError(apiextensionsv1beta1.AddToScheme(mgr.GetScheme()), "Cannot add API extensions to scheme")
+		kingpin.FatalIfError(stacks.Setup(mgr, log), "Cannot add stacks controllers to manager")
 
 		if *supportTemplates {
-			log.Info("Adding template controllers")
-			kingpin.FatalIfError(stacksTemplateControllerSetupWithManager(mgr), "Cannot add template controllers to manager")
+			kingpin.FatalIfError(templates.Setup(mgr, log), "Cannot add template controllers to manager")
 		}
 
-		// Start the Cmd
-		log.Info("Starting the manager")
-		kingpin.FatalIfError(mgr.Start(signals.SetupSignalHandler()), "Cannot start controller")
+		kingpin.FatalIfError(mgr.Start(ctrl.SetupSignalHandler()), "Cannot start controller manager")
+
 	case extUnpackCmd.FullCommand():
-		var outFile io.StringWriter
-		// stack unpack command was called, run the stack unpacking logic
-		if extUnpackOutfile == nil || *extUnpackOutfile == "" {
-			outFile = os.Stdout
-		} else {
-			openFile, err := os.Create(*extUnpackOutfile)
-			kingpin.FatalIfError(err, "Cannot create outfile")
-			defer closeOrError(openFile)
-			outFile = openFile
+		log := logging.NewLogrLogger(zl.WithName("stacks"))
+
+		outFile := os.Stdout
+		if *extUnpackOutfile != "" {
```

Comment:
It looks like this flag is not required, do we need to check it nil?

Reply:
I'm fairly certain kingpin ensures that string flags are always non-nil. If I add:

```go
		fmt.Printf("extUnpackOutfile value: '%s'\n", *extUnpackOutfile)
```

I don't get a nil pointer issue:

```
$ go run cmd/crossplane/main.go stack unstack --content-dir /tmp
extUnpackOutfile value: ''
{"level":"info","ts":1580721234.870985,"logger":"stacks","msg":"Unpacking stack","to":"/dev/stdout"}
main: error: failed to unpack stacks: Stack does not contain an app.yaml file
exit status 1
```

Reply:
Gotcha! Thanks for clarifying 🙂